File: puzzles/cpp/super-computer.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

int main() {
  vector<pair<int, int>> tasks;
  int n;
  cin >> n; cin.ignore();
  for (int i = 0; i < n; i++) {
    int startDay, duration;
    cin >> startDay >> duration; cin.ignore();
    pair<int, int> task = make_pair(startDay + duration - 1, startDay);
    tasks.push_back(task);
  }

  sort(tasks.begin(), tasks.end());

  vector<pair<int, int>> carryList;
  carryList.push_back(tasks[0]);
  int prevIndex = 0;
  for (int i = 0; i < n; i++) {
    pair<int, int> task = tasks[i];
    pair<int, int> prevTask = tasks[prevIndex];
    if (task.second > prevTask.first) {
      carryList.push_back(task);
      prevIndex = i;
    }
  }

  cout << carryList.size() << endl;
}

File: puzzles/js/super-computer.js
let reservationList = [];
const NB_CALCULATIONS = parseInt(readline());
for (let i = 0; i < NB_CALCULATIONS; ++i) {
  let inputs = readline().split(' ');
  let startDay = parseInt(inputs[0]);
  let duration = parseInt(inputs[1]);
  let task = {
    "startDay": startDay,
    "endDay": startDay + duration - 1
  }
  reservationList.push(task);
}

reservationList.sort(function(task1, task2) {
  return task1.endDay - task2.endDay;
});

let nbTasksCompleted = 1;
let prevIndex = 0;
for (let i = 1; i < NB_CALCULATIONS; ++i) {
  let task = reservationList[i];
  let prevTask = reservationList[prevIndex];
  // check if task doesn't overlap
  if (task.startDay > prevTask.endDay) {
    ++nbTasksCompleted;
    prevIndex = i;
  }
}

print(nbTasksCompleted);
